Adds should-publish-library-artifacts command

oguzkocer opened the pull request:
This PR adds `should-publish-library-artifacts` which checks the tag, branch name, sha1 and the pull request url to see if we'd want to publish the artifacts to S3. It's a companion to `publish-to-s3` command from https://github.com/wordpress-mobile/circleci-orbs/pull/72 and can be used before the code is checked out. The current behavior is, we'll continue to build if:
* It's from a tag
* It's from a PR
* It's from develop or trunk
* Fails if sha1 is missing
* Otherwise uses `circleci-agent step halt` to gracefully stop the build

**Note that this PR is targeting the `publish-to-s3` branch since it has commits from that PR. We can either merge #72 first and update the base branch to `develop` or merge this PR to #72 and merge them together to `develop`.**

Below script can be used to test various scenarios:

```sh
#!/bin/bash

CIRCLE_SHA1="sha1"

CIRCLE_TAG="tag"
CIRCLE_BRANCH="develop"
CIRCLE_PULL_REQUEST="pr"

if [[ -z "$CIRCLE_SHA1" ]]; then
    echo "Commit hash should always be available. Since we rely on this assumption in later commands, this acts as a sanity check"
    exit 1
fi

if [[ -n "$CIRCLE_TAG" ]]; then
    echo "Proceed because we are building a tag..."
elif [[ -n "$CIRCLE_PULL_REQUEST" ]]; then
    echo "Proceed because we are building a PR..."
elif [[ "$CIRCLE_BRANCH" == "develop" || "$CIRCLE_BRANCH" == "trunk" ]]; then
    echo "Proceed because we are building $CIRCLE_BRANCH..."
else
    echo "We only publish artifacts if the build is from a tag, a PR or from develop and trunk branches."
    echo "Skipping publishing artifacts!"
    # circleci-agent step halt
fi
```

oguzkocer committed:
Adds should-publish-library-artifacts command

mokagio commented on src/android/orb.yml:
```diff
@@ -218,4 +218,26 @@ commands:
+          name: Check if library artifacts should be published
+          command: |
+            if [[ -z "$CIRCLE_SHA1" ]]; then
+                echo "Commit hash should always be available. Since we rely on this assumption in later commands, this acts as a sanity check"
```
```suggestion
                echo "Commit hash should always be available. Since we rely on this assumption in later commands, this acts as a precondition check"
```

Rationale: it makes the message [more inclusive](https://english.stackexchange.com/questions/282282/near-universally-applicable-alternative-to-sanity-check).

mokagio commented:
> We can either merge #72 first and update the base branch to develop or merge this PR to #72 and merge them together to develop.

Whichever is easier for you, I'd say.

mokagio committed:
Add debug information to should-publish-library-artifacts

mokagio committed:
Add tip to halt message in should-publish-library-artifacts

It can happen that a developer pushes a branch for a finished feature
and then opens a PR for it.

In such case, depending on the load on the CircleCI queue, the build
might start before the PR is open, which would result in the S3
publishing halting, as intended.

There's not much we can do from a tooling perspective, short of building
and upload every source code check in, to work around this issue.

This commit adds what is hopefully going to be useful information for
developers confused by this behavior.

AliSoftware commented on src/android/orb.yml:
```diff
@@ -218,4 +218,26 @@ commands:
             [[ -n "$CIRCLE_BRANCH" ]] && PARAMS+=(--branch-name=$CIRCLE_BRANCH)
             [[ -n "$CIRCLE_PULL_REQUEST" ]] && PARAMS+=(--pull-request-url=$CIRCLE_PULL_REQUEST)
             ./gradlew --stacktrace :<<parameters.module_name>>:publishToS3 "${PARAMS[@]}"
+  should-publish-library-artifacts:
```
nit: this was unclear to me when I saw this command's name at call site in [this PR](https://github.com/wordpress-mobile/WordPress-Utils-Android/pull/73/files#diff-78a8a19706dbd2a4425dd72bdab0502ed7a2cef16365ab7030a5a0588927bf47R44) that the command behind this would be halting the rest of the workflow. Only when I saw the impl below with `step halt` did I get the intent.

What about renaming this to `continue-only-if-publish-library-artefacts` or `check-precondition-for-publish-artefacts`? (the term "precondition" to me kinda hints that it would halt if the precondition fails)

---

Also, I'm not sure of the benefit of having a command for that, as opposed to let the client decide, and make the invocation of that command in a client's job be conditional in the `config.yml` file of the host repo itself? (i.e. like [here](https://github.com/wordpress-mobile/WordPress-Utils-Android/pull/73/files#diff-78a8a19706dbd2a4425dd72bdab0502ed7a2cef16365ab7030a5a0588927bf47R44))

oguzkocer replied:
I am happy to rename this to `check-precondition-for-publish-artifacts` 👍 

---

> Also, I'm not sure of the benefit of having a command for that, as opposed to let the client decide, and make the invocation of that command in a client's job be conditional in the `config.yml` file of the host repo itself?

@AliSoftware I am sorry, I didn't understand this part. Could you please explain further?

For context, I wanted to use filters instead of a command such as this, but I couldn't find a way to get the behavior we want just using CircleCI filters. @jkmassel also took a look at it and the closest solution we both could find was to enable `only build pull requests` feature in CircleCI which meant that we wouldn't be building `trunk`. He also pointed out, and I agree, that we don't want to enable this option for every library, so it left us to do a command such as this one.

I'd like to get this filter in because some of the libraries (such as gutenberg-mobile) has a very costly setup (~20 min iirc) to get to a stage where we could publish the artifacts.

If you have any suggestions to address these concerns, I'd love to try them, because I am definitely not happy with this solution.

mokagio replied:
> I'd like to get this filter in because some of the libraries (such as gutenberg-mobile) has a very costly setup (~20 min iirc) to get to a stage where we could publish the artifacts.

+1 for @oguzkocer's rationale. I found the exact same issue when working with Bintray. Hopefully a CI provider with more advanced configuration abilities than CircleCI will help tidying up this process. In the meantime, having it as a dedicated, centralized step makes the CI workflow progression clearer when looking at it in CircleCI.

AliSoftware replied:
Thanks, I was lacking that context, makes more sense now that I have it with your explanation above.

Worth adding that feature request to the RFC p2 post about our future CI provider hunt then!!

oguzkocer replied:
> Thanks, I was lacking that context, makes more sense now that I have it with your explanation above.

👍 

> Worth adding that feature request to the RFC p2 post about our future CI provider hunt then!!

For sure - I plan on posting a few things like this into discussion post. In the meantime, I've chatted about these with @jkmassel and it looks like our likely choice will address them all 🥳

oguzkocer committed:
Better wording for sha1 check in should-publish-library-artifacts

Co-authored-by: Gio Lodi <giovanni.lodi42@gmail.com>

oguzkocer committed:
Merge pull request #74 from wordpress-mobile/add-debug-info-to-should-publish-library-artifacts

Add tip to halt message in should-publish-library-artifacts

oguzkocer commented:
> Base automatically changed from publish-to-s3 to trunk 5 minutes ago

That's a nice little future Github 👏

oguzkocer committed:
Rename should-publish-library-artifacts to check-precondition-for-publish-artifacts
